public/app/core/context/ModalsContextProvider.tsx

```diff
@@ -0,0 +1,114 @@
+import React, { useEffect, useState } from 'react';
+
+import { textUtil } from '@grafana/data';
+import { locationService } from '@grafana/runtime';
+import { ConfirmModal, ConfirmModalProps, ModalsContext } from '@grafana/ui';
+import { ModalsContextState } from '@grafana/ui/src/components/Modal/ModalsContext';
+import { ShowConfirmModalEvent, ShowModalReactEvent } from 'app/types/events';
+
+import appEvents from '../app_events';
+
+export interface Props {
+  children: React.ReactNode;
+}
+
+/**
+ * Implements the ModalsContext state logic (not used that much, only needed in edge cases)
+ * Also implements the handling of the events ShowModalReactEvent and ShowConfirmModalEvent.
+ */
+export function ModalsContextProvider(props: Props) {
+  const [state, setState] = useState<ModalsContextState>({
+    component: null,
+    props: {},
+    showModal: (component: React.ComponentType<any>, props: any) => {
+      setState({ ...state, component, props });
+    },
+    hideModal: () => {
+      setState({ ...state, component: null, props: {} });
+    },
```

Comment:
Are you able to explain why showModal/hideModal are attributes of the state here? This feels pretty unexpected to me.

Reply:
Ahh I see now, I think - it's used as a bit of a shortcut to construct the context object. I think you kind of pay the 'cost' of this with a bunch of bigger-than-nessicary object spreads later on whenever you update the state. I think this would be better as something like:

```ts
const [modalState, setModalState] = useState<{ component: React.ComponentType, props: ... }>();

// ...

const contextValue =  useMemo(() => {
  return {
    component: modalState.component,
    props: modalState.props,
    showModal: (...) => ...,
    showModal: (...) => ...
  }
}, [modalState])
```

You could probably even not have to store isOpen / onDismiss in state if you add them to the context value in the memo.

Reply:
@joshhunt thanks! that is cleaner!  will update